In [2]:
import numpy as np
import pandas as pd
import sklearn.datasets
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV

In [5]:
pd.set_option('display.max_columns', None)

In [6]:
# Working with dataset
breast_cancer_dataset = sklearn.datasets.load_breast_cancer()
print(breast_cancer_dataset)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

In [7]:
data_frames = pd.DataFrame(breast_cancer_dataset.data, columns= breast_cancer_dataset.feature_names)
data_frames.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  radius error  texture error  perimeter error  \
0                 0.07871        1.0950         0.9053            8.589   
1                 0.05667        0.5435         0.7339            3.398   
2                 0.05999        0.7456         0.7869            4.585   
3                 0.09744        0.4956         1.1560            3.445   
4                 0.05883        0.7572         0.7813            5.438   

   area error  smoothness error  compactness error  concavity error  \
0      153.40          0.006399            0.04904          0.05373   
1       74.08          0.005225            0.01308          0.01860   
2       94.03          0.006150            0.04006          0.03832   
3       27.23          0.009110            0.07458          0.05661   
4       94.44          0.011490            0.02461          0.05688   

   concave points error  symmetry error  fractal dimension error  \
0               0.01587         0.03003                 0.006193   
1               0.01340         0.01389                 0.003532   
2               0.02058         0.02250                 0.004571   
3               0.01867         0.05963                 0.009208   
4               0.01885         0.01756                 0.005115   

   worst radius  worst texture  worst perimeter  worst area  worst smoothness  \
0         25.38          17.33           184.60      2019.0            0.1622   
1         24.99          23.41           158.80      1956.0            0.1238   
2         23.57          25.53           152.50      1709.0            0.1444   
3         14.91          26.50            98.87       567.7            0.2098   
4         22.54          16.67           152.20      1575.0            0.1374   

   worst compactness  worst concavity  worst concave points  worst symmetry  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   worst fractal dimension  
0                  0.11890  
1                  0.08902  
2                  0.08758  
3                  0.17300  
4                  0.07678

In [8]:
data_frames['Label'] = breast_cancer_dataset.target
data_frames.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  radius error  texture error  perimeter error  \
0                 0.07871        1.0950         0.9053            8.589   
1                 0.05667        0.5435         0.7339            3.398   
2                 0.05999        0.7456         0.7869            4.585   
3                 0.09744        0.4956         1.1560            3.445   
4                 0.05883        0.7572         0.7813            5.438   

   area error  smoothness error  compactness error  concavity error  \
0      153.40          0.006399            0.04904          0.05373   
1       74.08          0.005225            0.01308          0.01860   
2       94.03          0.006150            0.04006          0.03832   
3       27.23          0.009110            0.07458          0.05661   
4       94.44          0.011490            0.02461          0.05688   

   concave points error  symmetry error  fractal dimension error  \
0               0.01587         0.03003                 0.006193   
1               0.01340         0.01389                 0.003532   
2               0.02058         0.02250                 0.004571   
3               0.01867         0.05963                 0.009208   
4               0.01885         0.01756                 0.005115   

   worst radius  worst texture  worst perimeter  worst area  worst smoothness  \
0         25.38          17.33           184.60      2019.0            0.1622   
1         24.99          23.41           158.80      1956.0            0.1238   
2         23.57          25.53           152.50      1709.0            0.1444   
3         14.91          26.50            98.87       567.7            0.2098   
4         22.54          16.67           152.20      1575.0            0.1374   

   worst compactness  worst concavity  worst concave points  worst symmetry  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   worst fractal dimension  Label  
0                  0.11890      0  
1                  0.08902      0  
2                  0.08758      0  
3                  0.17300      0  
4                  0.07678      0

In [9]:
data_frames.shape

(569, 31)

In [10]:
data_frames.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
Label                      0
dtype: int64

In [12]:
data_frames['Label'].value_counts()

#1 --> Benign

#0 --> Malignant

1    357
0    212
Name: Label, dtype: int64

In [15]:
# Define the features var & target var
X = data_frames.drop(columns='Label', axis=1)
Y = data_frames['Label']

In [16]:
print(X)
print(Y)

     mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mea

In [17]:
#Convert the target & feature in Numpy Arrays
X = np.asarray(X)
Y = np.asarray(Y)

In [21]:
print(X)
print(Y)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 

# GridSearchCV

In [22]:
# Selecting model
model = SVC()

In [23]:
# Define the Prarameters
parameters = {'kernel':['linear', 'poly', 'rbf', 'sigmoid'],
              'C':[1,5,10,20]}

In [24]:
#gridsearch 
classifier = GridSearchCV(model,parameters,cv=5)

In [25]:
#Fit the data X & Y
classifier.fit(X,Y)

GridSearchCV(cv=5, estimator=SVC(),
             param_grid={'C': [1, 5, 10, 20],
                         'kernel': ['linear', 'poly', 'rbf', 'sigmoid']})

In [26]:
classifier.cv_results_

{'mean_fit_time': array([4.39188008e+00, 1.16629124e-02, 1.83745384e-02, 2.90963173e-02,
        9.03080707e+00, 9.04250145e-03, 1.09869957e-02, 2.09493160e-02,
        1.36107263e+01, 1.11322403e-02, 9.79919434e-03, 2.44465351e-02,
        2.18056763e+01, 1.36934757e-02, 1.07444763e-02, 2.53839970e-02]),
 'std_fit_time': array([1.41045447e+00, 4.87758167e-03, 6.69808219e-03, 1.66147675e-02,
        2.01518235e+00, 1.94478126e-03, 2.52782105e-03, 2.76103243e-03,
        3.64571109e+00, 6.65027213e-04, 7.46723585e-04, 3.39549963e-03,
        7.01601896e+00, 2.68825535e-03, 1.31010089e-03, 7.65845804e-04]),
 'mean_score_time': array([0.00820484, 0.00340214, 0.01160555, 0.00699944, 0.00200129,
        0.00213079, 0.00600333, 0.00575743, 0.0016048 , 0.00338411,
        0.00560098, 0.00461783, 0.00342774, 0.00227132, 0.00462379,
        0.00566812]),
 'std_score_time': array([0.0102929 , 0.00080676, 0.00531556, 0.00244498, 0.00189846,
        0.00112633, 0.0008932 , 0.00138478, 0.0004934 , 

In [27]:
classifier.best_score_ # model gives best average best_accuracy

0.9525694767893185

In [28]:
classifier.best_params_ # model gives best parameters

{'C': 10, 'kernel': 'linear'}

In [30]:
classifier_dataframes = pd.DataFrame(classifier.cv_results_) # Converting into Dataframes a model hyperparamenter process
classifier_dataframes.head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
0       4.391880      1.410454         0.008205        0.010293       1   
1       0.011663      0.004878         0.003402        0.000807       1   
2       0.018375      0.006698         0.011606        0.005316       1   
3       0.029096      0.016615         0.006999        0.002445       1   
4       9.030807      2.015182         0.002001        0.001898       5   

  param_kernel                         params  split0_test_score  \
0       linear   {'C': 1, 'kernel': 'linear'}           0.947368   
1         poly     {'C': 1, 'kernel': 'poly'}           0.842105   
2          rbf      {'C': 1, 'kernel': 'rbf'}           0.850877   
3      sigmoid  {'C': 1, 'kernel': 'sigmoid'}           0.543860   
4       linear   {'C': 5, 'kernel': 'linear'}           0.947368   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.929825           0.973684           0.921053           0.955752   
1           0.885965           0.929825           0.947368           0.938053   
2           0.894737           0.929825           0.947368           0.938053   
3           0.456140           0.464912           0.385965           0.451327   
4           0.938596           0.973684           0.929825           0.964602   

   mean_test_score  std_test_score  rank_test_score  
0         0.945536        0.018689                4  
1         0.908663        0.039382               12  
2         0.912172        0.035444               11  
3         0.460441        0.050253               13  
4         0.950815        0.016216                2

In [31]:
grid_search_result = classifier_dataframes[['param_C','param_kernel','mean_test_score']]
grid_search_result

param_C param_kernel  mean_test_score
0        1       linear         0.945536
1        1         poly         0.908663
2        1          rbf         0.912172
3        1      sigmoid         0.460441
4        5       linear         0.950815
5        5         poly         0.922729
6        5          rbf         0.931501
7        5      sigmoid         0.411178
8       10       linear         0.952569
9       10         poly         0.920975
10      10          rbf         0.922714
11      10      sigmoid         0.402391
12      20       linear         0.949061
13      20         poly         0.919221
14      20          rbf         0.920944
15      20      sigmoid         0.398867

# Final Result

# ----------------------------------------------------------------

## In GridSearchCV

# ----------------------------------------------------------------

### Best Model Accueracy is 0.9525694767893185
### Best Model Parameters is {'C': 10, 'kernel': 'linear'}


# RandomSearchCV

In [41]:
# Selecting model
model = SVC()

In [42]:
# Define the Prarameters
parameters = {'kernel':['linear', 'poly', 'rbf', 'sigmoid'],
              'C':[1,5,10,20]}

In [43]:
#gridsearch 
random_classifier = RandomizedSearchCV(model,parameters,cv=5)

In [44]:
#Fit the data X & Y
random_classifier.fit(X,Y)

RandomizedSearchCV(cv=5, estimator=SVC(),
                   param_distributions={'C': [1, 5, 10, 20],
                                        'kernel': ['linear', 'poly', 'rbf',
                                                   'sigmoid']})

In [45]:
random_classifier.cv_results_

{'mean_fit_time': array([0.01308131, 0.01230073, 0.02117381, 0.00757236, 0.01883173,
        0.00873399, 0.00650048, 0.00982747, 4.85468912, 0.00926771]),
 'std_fit_time': array([2.58072868e-03, 3.20688500e-03, 3.14181400e-03, 2.13999047e-03,
        1.65979618e-03, 2.11484597e-03, 1.99601219e-03, 7.91813918e-04,
        2.03381296e+00, 2.04332971e-03]),
 'mean_score_time': array([0.00220132, 0.00219178, 0.00444183, 0.00496712, 0.00403056,
        0.00560088, 0.00299878, 0.00560145, 0.00270367, 0.00287213]),
 'std_score_time': array([1.16377818e-03, 1.16062122e-03, 1.12526732e-03, 6.30362852e-05,
        2.01581530e-03, 8.03226120e-04, 2.44849512e-03, 7.93567642e-04,
        9.87976294e-04, 2.44520775e-03]),
 'param_kernel': masked_array(data=['poly', 'poly', 'sigmoid', 'rbf', 'sigmoid', 'rbf',
                    'poly', 'rbf', 'linear', 'poly'],
              mask=[False, False, False, False, False, False, False, False,
                    False, False],
        fill_value='?',
     

In [48]:
random_classifier.best_score_ # model gives best average best_accuracy

0.9455364073901569

In [49]:
random_classifier.best_params_ # model gives best parameters

{'kernel': 'linear', 'C': 1}

In [50]:
random_classifier_dataframes = pd.DataFrame(classifier.cv_results_) # Converting into Dataframes a model hyperparamenter process
random_classifier_dataframes.head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_kernel  \
0       0.011001      0.001990         0.002998        0.002448         poly   
1       0.023637      0.003690         0.004999        0.000006      sigmoid   
2       9.125352      1.909272         0.002126        0.000457       linear   
3      20.971017      7.123508         0.002001        0.001266       linear   
4       0.009749      0.000389         0.006292        0.001438          rbf   

  param_C                         params  split0_test_score  \
0      20    {'kernel': 'poly', 'C': 20}           0.877193   
1       1  {'kernel': 'sigmoid', 'C': 1}           0.543860   
2       5   {'kernel': 'linear', 'C': 5}           0.947368   
3      20  {'kernel': 'linear', 'C': 20}           0.929825   
4       5      {'kernel': 'rbf', 'C': 5}           0.885965   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.921053           0.903509           0.938596           0.955752   
1           0.456140           0.464912           0.385965           0.451327   
2           0.938596           0.973684           0.929825           0.964602   
3           0.956140           0.956140           0.938596           0.964602   
4           0.929825           0.938596           0.947368           0.955752   

   mean_test_score  std_test_score  rank_test_score  
0         0.919221        0.027300                7  
1         0.460441        0.050253               10  
2         0.950815        0.016216                2  
3         0.949061        0.012816                3  
4         0.931501        0.024358                4

In [51]:
random_classifier_result = classifier_dataframes[['param_C','param_kernel','mean_test_score']]
random_classifier_result

param_C param_kernel  mean_test_score
0        1       linear         0.945536
1        1         poly         0.908663
2        1          rbf         0.912172
3        1      sigmoid         0.460441
4        5       linear         0.950815
5        5         poly         0.922729
6        5          rbf         0.931501
7        5      sigmoid         0.411178
8       10       linear         0.952569
9       10         poly         0.920975
10      10          rbf         0.922714
11      10      sigmoid         0.402391
12      20       linear         0.949061
13      20         poly         0.919221
14      20          rbf         0.920944
15      20      sigmoid         0.398867

# Final Result

# ----------------------------------------------------------------

## In RandomSearchCV

# ----------------------------------------------------------------

### Best Model Accueracy is 0.9455364073901569
### Best Model Parameters is {'kernel': 'linear', 'C': 1}